library/core/src/iter/traits/collect.rs

```diff
@@ -236,6 +236,49 @@ pub trait FromIterator<A>: Sized {
 /// ```
 #[rustc_diagnostic_item = "IntoIterator"]
 #[rustc_skip_array_during_method_dispatch]
+#[rustc_on_unimplemented(
```

Comment:
are these copied from iterator or is there a difference? Maybe document that with a normal comment

Reply:
Yes, I took them straight from Iterator and Future. I can iterate on whether removing the others is needed or if only a subset of conditions are needed. I strongly suspect we need to keep all on Iterator and only need a subset on IntoIterator

Reply:
So I took a quick look and funnily enough this change actually improves the situation of not knowing when the error is coming from an expression desugaring or type context, because all desugaring uses the `Into*` traits, while most normal code uses the actual traits, so we can leave the expression level suggestions in the conversion traits (almost, deref would need to be accounted, as per the `FIXME` comment) exclusively.